### ISSUE TITLE
Rationalize Frontier compiler entries between SCREAM and E3SM.

### ISSUE THREAD

**Issue by rljacob** [E3SM-Project/E3SM#6773]
Opening this issue to discuss how to make sure all scream cases build with the "frontier" machine description and at least one compiler entry so we can remove "frontier-scream-gpu" from config_machines.xml

See https://acme-climate.atlassian.net/wiki/spaces/EIDMG/pages/3446079573/How+to+describe+heterogenous+node+machines+with+CIME for background.

In https://github.com/E3SM-Project/scream/pull/2969#issuecomment-2311244667 it was noted that screams "craygnuamdgpu” tells the user that it "Uses Cray wrappers with Gnu compilers for the host, and uses the AMD Hip compiler for the GPU".  We should follow that convention.

**Comment by bartgol** [comment]
Cosmetic comment: I personally find the string `craygnuamdgpu` hard to parse. For "combo" compilers such as this, it may be be more readable to use `hostcompiler-devicecompiler`, that is add some dashes in there. In this case, `craygnu-amdgpu`, so that the mach+compiler combo would be `frontier_craygnu-amdgpu`.

**Comment by grnydawn** [comment]
To drive the merging of Frontier compiler entries, I think it would be useful to choose one or two Scream test cases within the E3SM machine/compiler configurations and resolve any issues encountered during their build and execution on Frontier. If necessary, we may create a new Scream test case.

**Comment by rljacob** [comment]
Use any of the scream test cases that already exist such as ne30_ne30.F2010-SCREAMv1 which is in e3sm_scream_v1_medres in https://github.com/E3SM-Project/E3SM/blob/49fdbe3661f2b8c95d8459081f500fae3a069ba0/cime_config/tests.py#L704C6-L704C28

Those all pass on frontier when using --machine frontier-scream-gpu --compiler craygnuamdgpu

You could start by just copying the craygnuamdgpu compiler entry to the frontier machine file while we figure out how to name these.

**Comment by rljacob** [comment]
During the EAMxx dev call, we decided on using a "-" between the cpu and gpu instead of putting "gpu" in the compiler name.   If there is no dash, that means its a cpu-only compile.
craycray-amd  = cray wrapper around cray compiler for host,  amd compiler on gpu
craygnu-amd = cray wrapper around gnu compiler, amd compiler on gpu
gnu-amd = use gfortran directly on host, amd compiler on gpu.

**Comment by bartgol** [comment]
@rljacob what if we used openacc? Should the compiler be called gnu-openacc? gnu-gnu? And what if we used openacc for f90, and nvcc for C++? Should we call it gnu-nvcc? gnu-openacc-nvcc?

**Comment by bartgol** [comment]
I'm just trying to understand what happens if we use nvcc for C++ and openacc for f90, both running on device. NVCC is a compiler, not a prog model. So would you do `gnu-gnu-nvcc`, since you use two different GPU compilers?

**Comment by grnydawn** [comment]
I am working on renaming the `craygnuamdgpu` and `crayclang-scream` compilers in the `frontier-scream-gpu` machine based on the basic approach Rob explained and discussions in the issue threads. The following summarizes the issues I encountered during the renaming process:

### 1. Compiler Wrapper
- Not all compiler configurations use a compiler wrapper. For example, `craygnuamdgpu` uses the `hipcc` compiler as `MPICXX`.

### 2. Programming Language
- Fortran compiler configurations differ from C++ compiler configurations. For example, `MPIFC` uses `ftn`, whereas `MPICXX` uses `hipcc` or `mpicxx`.

### 3. GPU Programming Framework
- There is no way to express the use of CUDA, HIP, OpenACC, or OpenMP offload.

### 4. Model-Specific Compiler
- The `craygnuamdgpu` compiler sets `MPICXX` as `hipcc`, which causes compiler errors on Omega because `hipcc` is not an MPI-enabled compiler.

---

Tentatively, I renamed `craygnuamdgpu` to `craygnu-hipcc` and `crayclang-scream` to `craycray-hipcc`. However, these new names still retain the issues listed above.

### Thoughts on Renaming
1. A compiler wrapper may not be necessary in the CIME compiler name. Compiler wrappers are usually not an issue during compilation, and in most cases, we know what compiler wrapper is used based on the system name.
2. We may want to distinguish the Fortran compiler and the C++ compiler because they might be different, such as `crayftn` for Fortran and `hipcc` for C++.
3. The C++ compiler name provides hints about the GPU programming framework in use, such as `nvcc` or `hipcc`.
4. We may still need to express additional information, such as model-specific details or the GPU programming framework.

### Proposed Naming Convention
Considering these factors, I came up with the following naming convention:

**\<Fortran compiler name without compiler wrapper>-\<C++(GPU) compiler name without compiler wrapper>-\<Optional specific information such as model name or GPU programming framework>**

For example:
- Rename `crayclang-scream` to **`cray-hipcc`**.
- Rename `craygnuamdgpu` to **`gnu-hipcc-eamxx`**. 
- Rename `crayclang` for cpu to **`cray`** 

The `eamxx` addition in the second name reflects that the compiler configuration uses `hipcc` for `MPICXX` and is likely incompatible with other C++ models, such as Omega.

**Comment by ndkeen** [comment]
Again for the record, I don't think this is the best approach.  We don't need those longer names.

**Comment by grnydawn** [comment]
> This is good but I would like to do away with "model-specific" compiler configs. Why does EAMxx need that?

Most compiler names might not include the third, model-specific part. The model name in "gnu-hipcc-eamxx" is added because it may not be able to compile the Omega code due to the "eamxx-specific" compiler configuration explained above. If we still want to remove the third part, I believe the "eamxx-specific" compiler configuration might need to be removed.

**Comment by grnydawn** [comment]
> Again for the record, I don't think this is the best approach. We don't need those longer names.

As noted above, most compiler names might not include the third, model-specific part. So, in most cases, the length of this naming scheme would be similar to other suggestions. I think we need to decide which information to retain in the compiler name and which to discard.

**Comment by rljacob** [comment]
The defining feature of gnu-hipcc-eamxx is that it doesn't include MPI.  So we should name it something like gnu-hipccnoMPI.  We want to concentrate on what the compiler is doing rather then what models wants it.

**Comment by grnydawn** [comment]
> The defining feature of gnu-hipcc-eamxx is that it doesn't include MPI. So we should name it something like gnu-hipccnoMPI. We want to concentrate on what the compiler is doing rather then what models wants it.

I agree with that. In fact, the latest renaming scheme I presented at today's Eamxx dev meeting supports `craygnu-hipcc` because it does not have a compiler wrapper. Therefore, we do not need the third part in the compiler name or `noMPI`.

**Comment by grnydawn** [comment]
This is a copy of the renaming option that I presented in EAMxx dev meeting:

```
[Fortran wrapper]<Fortran CPU compiler>[-[C++ wrapper]C++ DEVICE compiler]
```

**Exampes**
* crayclang-scream -> craycray-mpicxxhipcc
* craygnuamdgpu -> craygnu-hipcc
* crayclang -> craycray
* amdclanggpu -> crayamd-mpicxxhipcc

**Comment by rljacob** [comment]
How about instead of "mpicxxhipcc" just "mphipcc"   So   craycray-mphipcc  would be the name.

**Comment by grnydawn** [comment]
> How about instead of "mpicxxhipcc" just "mphipcc" So craycray-mphipcc would be the name instead of "craycray-mpicxxhipcc"

@rljacob , I like the shortened name "craycray-mphipcc." I will change the names of the other compilers accordingly for AMD and GNU compilers.

**Comment by grnydawn** [comment]
I think that PR #6990 and other following PRs have addressed this issue. Closing the issue.
